# Module 1 - Thu thập dữ liệu từ [nhatot.vn](https://www.nhatot.com/thue-can-ho-chung-cu-tp-ho-chi-minh)

## 1. Mục tiêu

### 1.1. Kịch bản

Thực hiện thu thập dữ liệu chung cư cho thuê ở thành phố Hồ Chí Minh từ [nhatot.vn](https://www.nhatot.com/thue-can-ho-chung-cu-tp-ho-chi-minh), sau khi tìm hiểu và hỏi bạn bè thì tìm được API của chợ tốt, từ đó việc crawl trở nên dễ dàng.

Trong source page có một script tag với id `__NEXT_DATA__` chứ thông tin query của page tiếp theo, dựa vào đó ta có đoán các query param:
```js
        "nextQuery": {
          "queryFilter": "region_v2=13000\u0026cg=1010\u0026o=40\u0026page=3\u0026st=u,h\u0026limit=20\u0026w=1",
          "preAsPath": "/thue-can-ho-chung-cu-tp-ho-chi-minh?page=3",
          "page": "3",
          "standardParam": {
            "region_v2": "13000",
            "cg": 1010,
            "o": 40,
            "page": 3,
            "st": "u%2Ch",
            "limit": "20",
            "w": "1"
          }
        }
```


API url có dạng như sau: [https://gateway.chotot.com/v1/public/ad-listing?region_v2=13000&cg=1010&limit=1&o=0&st=u,h&page=0](https://gateway.chotot.com/v1/public/ad-listing?region_v2=13000&cg=1010&limit=1&o=0&st=u,h&page=0)

Trong đó:
+ `https://gateway.chotot.com/v1/public/ad-listing?` : API Endpoint
+ Các query param:
    + **region_v2**: mã TP/tỉnh VD: 13000 là HCM, 12000 là HN
    + **cg**: mã category VD 1010 tương ứng chung cư
    + **limit**: giới hạn số record (pagination)
    + **o**: bắt đầu từ record số mấy (pagination)
    + **page**: trang bao nhiêu (pagination)
    + **st**: không rõ lắm nhưng chắc là thứ tự sắp xếp


Dưới đây là response chứa thông tin một căn chung cư cho thuê

```json
{
    "variant": 0,
    "total": 58788,
    "ads": [
        {
            "ad_id": 137992545,
            "list_id": 99878550,
            "list_time": 1669997628437,
            "date": "2 phút trước",
            "account_id": 10849761,
            "account_oid": "4701b658e345b8bbc5c62500e4886893",
            "account_name": "Phạm Quốc Cường",
            "subject": "STUDIO FULL NT_30m2_gần Crescent Mall_Phú Mỹ Hưng",
            "body": "🌅𝐍𝐄𝐖 𝐍𝐄𝐖 𝐍𝐄𝐖 𝐍𝐄𝐖 𝐍𝐄𝐖 𝟏𝟎𝟎𝟎%🎉\n🍍𝗖𝗔̆𝗡 𝗛𝗢̣̂ 𝗠𝗢̛́𝗜 𝗫𝗔̂𝗬 𝗦𝗜𝗘̂𝗨 𝗨̛𝗨 𝗗𝗔̃𝗜✨\n\nVị trí:  Đường Nguyễn Lương Bằng , Tân Phú , Quận 7.\n\n👉🏻Phòng mới xây 100% đầy đủ nội thất chưa qua sử dụng✨\n👉Căn STUDIO - 1PN RIÊNG ở Được 3 - 4 người \n👉🏻 Giờ giấc tự do, bảo vệ 24/24 \n👉🏻 Cho nuôi thú cưng với cam kết giữ gìn vệ sinh và trật tự toà nhà\n👉🏻Diện tích rộng thoáng, cửa sổ lớn \n👉🏻Bếp từ, tivi, máy lạnh, kệ tủ, máy lạnh, giường nệm, bàn ghế ăn, v...v\n👉🏻Giặt sấy.\n👉🏻Thang máy\n\n📞 Call/zalo/mess Để được xem phòng FREE nhé !",
            "category": 1010,
            "category_name": "Căn hộ/Chung cư",
            "area": 102,
            "area_name": "Quận 7",
            "region": 13,
            "region_name": "Tp Hồ Chí Minh",
            "company_ad": true,
            "type": "u",
            "price": 5000000,
            "price_string": "5 triệu/tháng",
            "image": "https://cdn.chotot.com/8bkO4_wLHz9BRjJ0svZL6W7qnTQRPKQMGIAY_bEWhls/preset:listing/plain/4fb4350bebcba90d6ed4d16b9d5c6da1-2794066158885310428.jpg",
            "webp_image": "https://cdn.chotot.com/vg1EgIriCAE7dNIb7RJKH6BUY9D3xVagwirpCBr-8oY/preset:listing/plain/4fb4350bebcba90d6ed4d16b9d5c6da1-2794066158885310428.webp",
            "videos": [],
            "number_of_images": 12,
            "avatar": "https://cdn.chotot.com/uac2/10849761",
            "rooms": 1,
            "property_legal_document": 1,
            "apartment_type": 2,
            "region_v2": 13000,
            "area_v2": 13102,
            "ward": 9473,
            "ward_name": "Phường Tân Phú",
            "balconydirection": 8,
            "location": "10.72605,106.72377",
            "longitude": 106.72377,
            "latitude": 10.72605,
            "phone_hidden": true,
            "owner": false,
            "protection_entitlement": false,
            "escrow_can_deposit": 2,
            "zero_deposit": false,
            "pty_jupiter": "0"
        }
    ]
}
```

Kết quả trên chưa có diện tích và số phòng vệ sinh và một số thông tin khác, do đó cần request tới chi tiết của từng quảng cáo dựa vào `list_id`

API URL: [https://gateway.chotot.com/v1/public/ad-listing/99878550](https://gateway.chotot.com/v1/public/ad-listing/99878550) với 99878550 chính là `list_id` của tin quảng cáo

Kết quả như sau:

```json
{
    "ad": {
        "ad_id": 137992545,
        "list_id": 99878550,
        "list_time": 1669997628437,
        "date": "2 giờ trước",
        "account_id": 10849761,
        "account_oid": "4701b658e345b8bbc5c62500e4886893",
        "account_name": "Phạm Quốc Cường",
        "subject": "STUDIO FULL NT_30m2_gần Crescent Mall_Phú Mỹ Hưng",
        "body": "🌅𝐍𝐄𝐖 𝐍𝐄𝐖 𝐍𝐄𝐖 𝐍𝐄𝐖 𝐍𝐄𝐖 𝟏𝟎𝟎𝟎%🎉\n🍍𝗖𝗔̆𝗡 𝗛𝗢̣̂ 𝗠𝗢̛́𝗜 𝗫𝗔̂𝗬 𝗦𝗜𝗘̂𝗨 𝗨̛𝗨 𝗗𝗔̃𝗜✨\n\nVị trí:  Đường Nguyễn Lương Bằng , Tân Phú , Quận 7.\n\n👉🏻Phòng mới xây 100% đầy đủ nội thất chưa qua sử dụng✨\n👉Căn STUDIO - 1PN RIÊNG ở Được 3 - 4 người \n👉🏻 Giờ giấc tự do, bảo vệ 24/24 \n👉🏻 Cho nuôi thú cưng với cam kết giữ gìn vệ sinh và trật tự toà nhà\n👉🏻Diện tích rộng thoáng, cửa sổ lớn \n👉🏻Bếp từ, tivi, máy lạnh, kệ tủ, máy lạnh, giường nệm, bàn ghế ăn, v...v\n👉🏻Giặt sấy.\n👉🏻Thang máy\n\n📞 Call/zalo/mess Để được xem phòng FREE nhé !",
        "category": 1010,
        "category_name": "Căn hộ/Chung cư",
        "area": 102,
        "area_name": "Quận 7",
        "region": 13,
        "region_name": "Tp Hồ Chí Minh",
        "company_ad": true,
        "type": "u",
        "price": 5000000,
        "price_string": "5 triệu/tháng",
        "videos": [],
        "images": [
            "https://cdn.chotot.com/g87MLn8GdA_nVtkGkKTEgj5urwlr243rdaQCPqBx1-8/preset:view/plain/4fb4350bebcba90d6ed4d16b9d5c6da1-2794066158885310428.jpg",
            "https://cdn.chotot.com/fbD9vZqm4Ouw6Sje2sRfxZkZL8BxgvT-QlpRGylb8Is/preset:view/plain/58f7e6a1e5995e1068690fbb167b8f87-2794066158835523526.jpg",
            "https://cdn.chotot.com/HYVC9lvbYnc1-vzwnFMs13GIgnul01cHSHyzP3fszTg/preset:view/plain/90a5d2caf1fb65d3fbf029cb4c334e7f-2794066160764424156.jpg",
            "https://cdn.chotot.com/WnydTbTqXQ6O4dk3qkhsm5PHP0qzksQeNpHY7l-33Vs/preset:view/plain/1c4cd603294cae16c82d07e8e63e2f2e-2794066160698950809.jpg",
            "https://cdn.chotot.com/4PJmguzoe4ADwfcKj6e6QpASC9IlN0Xf11t4Lq9DJ6U/preset:view/plain/3fadf921a63a9ad4d6b78a8d1fa5f371-2794066162538000099.jpg",
            "https://cdn.chotot.com/wEqPnpVD0qe_i9qi6RnLV7ufK5p4xPosmFv31vau9ck/preset:view/plain/df51f1c70066ea4b1d9edc80691c3f11-2794066162909163743.jpg",
            "https://cdn.chotot.com/ELUFA0HOvakovRdz5QXd3PiBSsvwlAecoiz3IHFgaIs/preset:view/plain/d9eefb0757318e1128f1a17cc1cfdff8-2794066164449175680.jpg",
            "https://cdn.chotot.com/KdZvmNbD3nTxNrxb1FGu_TXQn1AzKiCCj8O6uTXT7R4/preset:view/plain/926b82e3b092f6abbf239dcfedb7f1ad-2794066164100411615.jpg",
            "https://cdn.chotot.com/3rf6zLzyvn43oin7WOx_gItzGDzHpMEJJ2ebxx_VY08/preset:view/plain/6d74f775e386998f463ed75a3648d487-2794066165316964988.jpg",
            "https://cdn.chotot.com/kHl2nGjsM-JuVrY159gOx0IXfPEKOwvMAiKzzrMwtS0/preset:view/plain/cb15184e06f30de32a3dc4959ca0ee7e-2794066165522839133.jpg",
            "https://cdn.chotot.com/1ky255uMNjPpSFPxMJaciWNA-xpBYr77wfY0LcIAL-0/preset:view/plain/8aef0fce4ce1bc981a618719db556ad7-2794066166508212860.jpg",
            "https://cdn.chotot.com/IKQcbqQD0rtt0CsIFtRdVoY_ysy25HDX8ApDHUZtUSY/preset:view/plain/bc4e16acc4f04c72e50a716dd60bdcb5-2794066166599107056.jpg"
        ],
        "number_of_images": 12,
        "avatar": "https://cdn.chotot.com/uac2/10849761",
        "rooms": 1,
        "property_legal_document": 1,
        "apartment_type": 2,
        "size": 30,
        "region_v2": 13000,
        "area_v2": 13102,
        "ward": 9473,
        "ward_name": "Phường Tân Phú",
        "toilets": 1,
        "balconydirection": 8,
        "pty_map": "https://cdn.chotot.com/admincentre/location/10x726850_106x724570_c.jpg",
        "pty_map_modifier": 0.0008,
        "deposit": 5000000,
        "contain_videos": 2,
        "location": "10.72605,106.72377",
        "longitude": 106.72377,
        "latitude": 10.72605,
        "phone": "090378****",
        "type_name": "Cho thuê",
        "reviewer_image": "https://static.chotot.com.vn/thumbs/admin/9999999611.png",
        "reviewer_nickname": "Hoàng Cường",
        "thumbnail_image": "https://cdn.chotot.com/8bkO4_wLHz9BRjJ0svZL6W7qnTQRPKQMGIAY_bEWhls/preset:listing/plain/4fb4350bebcba90d6ed4d16b9d5c6da1-2794066158885310428.jpg",
        "owner": false,
        "protection_entitlement": false,
        "escrow_can_deposit": 2,
        "zero_deposit": false,
        "street_name": "Nguyễn Lương Bằng",
        "pty_jupiter": "0"
    },
    "parameters": [
        {
            "id": "ward",
            "value": "Phường Tân Phú",
            "label": " Phường"
        },
        {
            "id": "size",
            "value": "30 m²",
            "label": "Diện tích"
        },
        {
            "id": "rooms",
            "value": "1 phòng",
            "label": "Số phòng ngủ"
        },
        {
            "id": "toilets",
            "value": "1 phòng",
            "label": "Số phòng vệ sinh"
        },
        {
            "id": "balconydirection",
            "value": "Tây Nam",
            "label": "Hướng ban công"
        },
        {
            "id": "apartment_type",
            "value": "Căn hộ dịch vụ",
            "label": "Loại hình căn hộ"
        },
        {
            "id": "property_legal_document",
            "value": "Đã có sổ",
            "label": "Giấy tờ pháp lý"
        },
        {
            "id": "area",
            "value": "Quận 7",
            "label": "Quận"
        },
        {
            "id": "region",
            "value": "Tp Hồ Chí Minh",
            "label": "Tỉnh"
        },
        {
            "id": "address",
            "value": "Nguyễn Lương Bằng, Phường Tân Phú, Quận 7, Tp Hồ Chí Minh",
            "label": "Địa chỉ"
        },
        {
            "id": "deposit",
            "value": "5.000.000 đ",
            "label": "Số tiền cọc"
        }
    ],
    "ad_params": {
        "address": {
            "id": "address",
            "value": "Nguyễn Lương Bằng, Phường Tân Phú, Quận 7, Tp Hồ Chí Minh",
            "label": "Địa chỉ"
        },
        "apartment_type": {
            "id": "apartment_type",
            "value": "Căn hộ dịch vụ",
            "label": "Loại hình căn hộ"
        },
        "area": {
            "id": "area",
            "value": "Quận 7",
            "label": "Quận"
        },
        "balconydirection": {
            "id": "balconydirection",
            "value": "Tây Nam",
            "label": "Hướng ban công"
        },
        "deposit": {
            "id": "deposit",
            "value": "5.000.000 đ",
            "label": "Số tiền cọc"
        },
        "property_legal_document": {
            "id": "property_legal_document",
            "value": "Đã có sổ",
            "label": "Giấy tờ pháp lý"
        },
        "region": {
            "id": "region",
            "value": "Tp Hồ Chí Minh",
            "label": "Tỉnh"
        },
        "rooms": {
            "id": "rooms",
            "value": "1 phòng",
            "label": "Số phòng ngủ"
        },
        "size": {
            "id": "size",
            "value": "30 m²",
            "label": "Diện tích"
        },
        "toilets": {
            "id": "toilets",
            "value": "1 phòng",
            "label": "Số phòng vệ sinh"
        },
        "ward": {
            "id": "ward",
            "value": "Phường Tân Phú",
            "label": " Phường"
        }
    }
}
```


| Column         | Description                                |
|----------------|--------------------------------------------|
| ad_id          | id của tin (để phân biệt)                  |
| list_id        | id trong danh sách ads (dùng để phân biệt) |
| list_time      | Thời gian đăng tin                         |
| price          | Giá thuê (VND)                             |
| price_string   | Giá thuê (triệu / tháng) cho dễ nhìn       |
| size           | Diện tích (m2)                             |
| rooms          | Số phòng                                   |
| toilets        | Số nhà vệ sinh                             |
| body           | Mô tả                                      |
| street_name    | Đường                                      |
| ward_name      | Phường                                     |
| area_name      | Quận huyện                                 |
| region_name    | Thành phố, tỉnh                            |
| zero_deposit   | Có cần tiền cọc không?                     |
| deposit        | Tiền cọc                                   |
| longitude      | Kinh độ                                    |
| latitude       | Vĩ độ                                      |
| owner          | Có phải chủ nhà không?                     |
| company_ad     | Tin của các nhân hay công ty               |
| apartment_type | Kiểu căn hộ / chung cư                     |


Trang web chỉ có 20000 tin tương ứng với 20000 record, hạn chế số record mỗi lần request sẽ là 100, vậy sẽ cần request `20000 / 100 = 200` lần để request lấy thông tin danh sách các căn hộ, rồi

## 2. Thực hiện

### 2.1. Import thư viện cần thiết

In [7]:
import csv
import time
import requests
import json

### 2.2. CONSTANT

In [8]:
BASE_URL = 'https://gateway.chotot.com/v1/public/ad-listing'

LIMIT = 100
LIMIT_PAGE = 200
SLEEP_TIME = 2
REGION_V2 = 13000
CG_APARTMENT = 1010

### 2.3. Lấy danh sách căn hộ chưa đầy đủ thông tin

Hàm lấy thông tin 100 record mỗi lượt

In [9]:
def get_data_per_page(page):
    url = f'{BASE_URL}?region_v2={REGION_V2}&cg={CG_APARTMENT}&limit={LIMIT}&o={LIMIT * page}&st=u,h&page={page}&w=1'
    res = requests.get(url)
    time.sleep(SLEEP_TIME)
    data = res.json()
    return data['ads']


Cách thực hiện: Crawl lần lượt 200 page, mỗi page 100 record

In [10]:
# get all apartments data
def get_data():
    all_apartments = []
    for i in range(LIMIT_PAGE):
        print(f'get data in page {i}')
        list_apartment = get_data_per_page(i)
        all_apartments.extend(list_apartment)

    return all_apartments

In [11]:
apartments = get_data()
apartments[:5]

get data in page 0
get data in page 1
get data in page 2
get data in page 3
get data in page 4
get data in page 5
get data in page 6
get data in page 7
get data in page 8
get data in page 9
get data in page 10
get data in page 11
get data in page 12
get data in page 13
get data in page 14
get data in page 15
get data in page 16
get data in page 17
get data in page 18
get data in page 19
get data in page 20
get data in page 21
get data in page 22
get data in page 23
get data in page 24
get data in page 25
get data in page 26
get data in page 27
get data in page 28
get data in page 29
get data in page 30
get data in page 31
get data in page 32
get data in page 33
get data in page 34
get data in page 35
get data in page 36
get data in page 37
get data in page 38
get data in page 39
get data in page 40
get data in page 41
get data in page 42
get data in page 43
get data in page 44
get data in page 45
get data in page 46
get data in page 47
get data in page 48
get data in page 49
get data i

[{'ad_id': 140195209,
  'list_id': 101689535,
  'list_time': 1670069218000,
  'date': '2 phút trước',
  'account_id': 23305211,
  'account_oid': 'eddceb994337d32eb38361ecec83f163',
  'account_name': 'Tạ Thanh Hải',
  'subject': 'Căn hộ Nội Thất Cao Cấp - Phường 15, Phú Nhuận',
  'body': 'Địa chỉ: Phan Đình Phùng, Q.Phú Nhuận\n\n📌Nội Thất: như hình \nTiện ích Toà nhà:\n+ Hầm xe\n+ Thang máy\n+ Cửa cuốn, hệ thống PCCC và camera an ninh\n+ Giặt sấy riêng trong \n+ Được nuôi thú cưng',
  'category': 1010,
  'category_name': 'Căn hộ/Chung cư',
  'area': 111,
  'area_name': 'Quận Phú Nhuận',
  'region': 13,
  'region_name': 'Tp Hồ Chí Minh',
  'company_ad': True,
  'type': 'u',
  'price': 12500000,
  'price_string': '12,5 triệu/tháng',
  'image': 'https://cdn.chotot.com/rgv865tEeWrguIOGnGWqPzNRnguAd55lf3nNJ83jTfY/preset:listing/plain/857384db2a324c03a9ec0ce3df9e1ef3-2801910873274840025.jpg',
  'webp_image': 'https://cdn.chotot.com/771IXNmr5hYEWpSs99gqHiZTBw6omelNfxBuN9OAJTA/preset:listing/pl

### 2.4. Lấy thông tin từng căn hộ

Lấy danh sách `list_id` để truy vấn từng căn hộ

In [35]:
hcm_apartments = list((filter(lambda apartment: apartment['region'] == 13 and apartment['category'] == 1010, apartments)))
ad_listing_id = list(set([apartment['list_id'] for apartment in hcm_apartments]))
len(ad_listing_id)

19913

Có chính xác 19913 tin cho thuê nhà ở thành phố Hồ Chí Minh, cần lấy thông tin từng căn hộ một

In [36]:
def get_apartment_detail(list_id):
    url = f'{BASE_URL}/{list_id}'
    res = requests.get(url)
    data = res.json()
    return data

In [23]:
def get_all_apartments():
    all_apartments = []
    for i, list_id in enumerate(ad_listing_id):
        print(f'get aparment #{i}, list_id: {list_id}')
        apartment = get_apartment_detail(list_id)
        all_apartments.append(apartment)
    return all_apartments

In [24]:
all_apartments = get_all_apartments()
all_apartments[:5]

get aparment #0, list_id: 101449743
get aparment #1, list_id: 101580825
get aparment #2, list_id: 101449753
get aparment #3, list_id: 101580839
get aparment #4, list_id: 101580840
get aparment #5, list_id: 101318703
get aparment #6, list_id: 101449793
get aparment #7, list_id: 93192262
get aparment #8, list_id: 100794446
get aparment #9, list_id: 100794451
get aparment #10, list_id: 101580898
get aparment #11, list_id: 101449831
get aparment #12, list_id: 101318771
get aparment #13, list_id: 92799106
get aparment #14, list_id: 100663434
get aparment #15, list_id: 101318806
get aparment #16, list_id: 101449893
get aparment #17, list_id: 101318829
get aparment #18, list_id: 101580975
get aparment #19, list_id: 101318842
get aparment #20, list_id: 101580999
get aparment #21, list_id: 101449928
get aparment #22, list_id: 101449929
get aparment #23, list_id: 101581032
get aparment #24, list_id: 101449966
get aparment #25, list_id: 101581047
get aparment #26, list_id: 101449975
get aparment 

[{'ad': {'ad_id': 139905225,
   'list_id': 101449743,
   'list_time': 1669427357000,
   'date': '1 tuần trước',
   'account_id': 12024575,
   'account_oid': '4e8066b4de746ddb72b9ba9145a5781a',
   'account_name': 'Anh Thu',
   'subject': 'Căn hộ CMT8 Q10, Cửa sổ, Máy giặt riêng, Gác',
   'body': 'Căn hộ Cách Mạng Tháng 8 Q10, Cửa sổ, Máy giặt riêng, Gác chỉ 9tr\n\nCăn hộ Cách Mạng Tháng 8 - Tô Hiến Thành - Sư Vạn Hạnh - 3/2 Q10 - Cửa sổ - Máy giặt riêng - Gác chỉ 9tr\nVị trí: Căn hộ Cách Mạng Tháng 8 Q10 gần cv Lê Thị Riêng, đh Bách Khoa, đh Huflit, bv 115\nMáy giặt riêng, hút mùi, bảo vệ 24/7, thang máy .\nĐiện 4k/kwh , nước 150 k / người\n2 người 2 xe 0 thú cưng\nLh: a.Thu',
   'category': 1010,
   'category_name': 'Căn hộ/Chung cư',
   'area': 105,
   'area_name': 'Quận 10',
   'region': 13,
   'region_name': 'Tp Hồ Chí Minh',
   'company_ad': True,
   'type': 'u',
   'price': 9000000,
   'price_string': '9 triệu/tháng',
   'videos': [],
   'images': ['https://cdn.chotot.com/LuJhr1G_

In [ ]:
# save to json file for backup
def save_to_json():
    with open('./data/apartments.json', 'w', encoding='utf-8') as f:
        json.dump(all_apartments, f, ensure_ascii=False, indent=4)

## 2.5. Filter data

In [37]:
need_keys = ['ad_id', 'list_id', 'list_time', 'subject', 'body', 'price', 'price_string', 'size', 'rooms', 'toilets', 'area_name', 'region_name', 'ward_name', 'street_name', 'deposit', 'zero_deposit', 'longitude', 'latitude', 'owner', 'company_ad', 'apartment_type']

def filter_apartment(apartment):
    new_apartment = { key: apartment['ad'][key] if key in apartment['ad'] else '' for key in need_keys }
    if apartment['ad_params']['apartment_type']['value']:
        new_apartment['apartment_type'] = apartment['ad_params']['apartment_type']['value']
    if new_apartment['deposit'] == '':
        new_apartment['deposit'] = 0
    return new_apartment

all_apartments1 = list(filter(lambda a: 'ad' in a,all_apartments))

new_apartments = list(map(filter_apartment, all_apartments1))
len(new_apartments)

19911

### 2.5. Write to csv

In [38]:
list_apartment = [[apartment[key] for key in need_keys] for apartment in new_apartments]

with open('./data/hcm-apartment-rent-data.csv', 'w') as f:
    write = csv.writer(f)
    write.writerow(need_keys)
    write.writerows(list_apartment)